File: framec_tests/golang/match/match_test.go
package match

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func (m *matchStruct) log(msg string) {
	m.tape = append(m.tape, msg)
}

/// Test matching the empty string.
/// TODO: Matching the empty string currently only works in multi-string
/// patterns. The pattern `//`, which should match only the empty string,
/// instead produces a parse error.

func TestEmptyString(t *testing.T) {
	sm := NewMatch()
	x := sm.(*matchStruct)
	sm.Empty()
	sm.Onstring("")
	assert.Equal(t, []string{"empty"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("hi")
	assert.Equal(t, []string{"?"}, x.tape)
}

func TestIntegerMatch(t *testing.T) {

	sm := NewMatch()
	x := sm.(*matchStruct)
	sm.Simple()
	sm.OnInt(0)
	assert.Equal(t, []string{"0"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(42)
	assert.Equal(t, []string{"42"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(-200)
	assert.Equal(t, []string{"-200"}, x.tape)
}

/* Test simple string matching.
TODO: Testing revealed some limitations:
 * Frame does not support UTF-8 graphemes larger than 1 byte, so we're
   restricted to ASCII.
 * Frame does not have a way to match the '/' or '|' characters,
   which are part of the matching syntax.
*/
func TestStringMatch(t *testing.T) {

	sm := NewMatch()
	x := sm.(*matchStruct)
	sm.Simple()
	sm.Onstring("hello")
	assert.Equal(t, []string{"hello"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("goodbye")
	assert.Equal(t, []string{"goodbye"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("Testing 1, 2, 3...")
	assert.Equal(t, []string{"testing"}, x.tape)
	assert.Equal(t, []string{"testing"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("$10!")
	assert.Equal(t, []string{"money"}, x.tape)
	assert.Equal(t, []string{"money"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("missing")
	assert.Equal(t, []string{"?"}, x.tape)
	assert.Equal(t, []string{"?"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("Testing")
	assert.Equal(t, []string{"?"}, x.tape)
	assert.Equal(t, []string{"?"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("")
	assert.Equal(t, []string{"?"}, x.tape)
	assert.Equal(t, []string{"?"}, x.tape)
}

/* Test the multiple match syntax for integers. */

func TestIntegerMultiMatch(t *testing.T) {
	sm := NewMatch()
	x := sm.(*matchStruct)
	sm.Multi()
	sm.OnInt(3)
	assert.Equal(t, []string{"3|-7"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(-7)
	assert.Equal(t, []string{"3|-7"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(-4)
	assert.Equal(t, []string{"-4|5|6"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(5)
	assert.Equal(t, []string{"-4|5|6"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(6)
	assert.Equal(t, []string{"-4|5|6"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(4)
	assert.Equal(t, []string{"?"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(0)
	assert.Equal(t, []string{"?"}, x.tape)

}

// Test the multiple match syntax for integers. Also tests matching
// whitespace-only strings.
func TestStringMultiMatch(t *testing.T) {
	sm := NewMatch()
	x := sm.(*matchStruct)
	sm.Multi()
	sm.Onstring("$10")
	assert.Equal(t, []string{"symbols"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("12.5%")
	assert.Equal(t, []string{"symbols"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("@#*!")
	assert.Equal(t, []string{"symbols"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring(" ")
	assert.Equal(t, []string{"whitespace"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("  ")
	assert.Equal(t, []string{"whitespace"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("\t")
	assert.Equal(t, []string{"whitespace"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("\n")
	assert.Equal(t, []string{"whitespace"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("10")
	assert.Equal(t, []string{"?"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("#")
	assert.Equal(t, []string{"?"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("   ")
	assert.Equal(t, []string{"?"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("")
	assert.Equal(t, []string{"?"}, x.tape)
	x.tape = x.tape[:0]
}

/* Test nested integer matching. */
func TestIntegerNestedMatch(t *testing.T) {
	sm := NewMatch()
	x := sm.(*matchStruct)
	sm.Nested()
	sm.OnInt(1)
	assert.Equal(t, []string{"1-3", "1"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(2)
	assert.Equal(t, []string{"1-3", "2"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(3)
	assert.Equal(t, []string{"1-3", "3"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(4)
	assert.Equal(t, []string{"4-5", "4"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(5)
	assert.Equal(t, []string{"4-5", "5"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(10)
	assert.Equal(t, []string{"too big"}, x.tape)
	x.tape = x.tape[:0]
	sm.OnInt(0)
	assert.Equal(t, []string{"too small"}, x.tape)

}

/* Test nested string matching. */
func TestStringNestedMatch(t *testing.T) {
	sm := NewMatch()
	x := sm.(*matchStruct)
	sm.Nested()
	sm.Onstring("hello")
	assert.Equal(t, []string{"greeting", "English"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("hola")
	assert.Equal(t, []string{"greeting", "Spanish"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("bonjour")
	assert.Equal(t, []string{"greeting", "French"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("goodbye")
	assert.Equal(t, []string{"farewell", "English"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("adios")
	assert.Equal(t, []string{"farewell", "Spanish"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("au revoir")
	assert.Equal(t, []string{"farewell", "French"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("hallo")
	assert.Equal(t, []string{"?"}, x.tape)
	x.tape = x.tape[:0]
	sm.Onstring("ciao")
	assert.Equal(t, []string{"?"}, x.tape)
}

/* Test hierarchical integer matching. */
func TestIntegerHierarchicalMatch(t *testing.T) {
	sm := NewMatch()
	x := sm.(*matchStruct)
	sm.Child()
	sm.OnInt(0)
	assert.Equal(t, MatchState_Final, x._compartment_.State)
	assert.Empty(t, x.tape)

	sm = NewMatch()
	x = sm.(*matchStruct)
	sm.Child()
	sm.OnInt(4)
	assert.Equal(t, MatchState_ChildMatch, x._compartment_.State)
	assert.Equal(t, []string{"4"}, x.tape)

	x.tape = x.tape[:0]
	sm.OnInt(5)
	assert.Equal(t, MatchState_Final, x._compartment_.State)
	assert.Equal(t, []string{"5"}, x.tape)

	sm = NewMatch()
	x = sm.(*matchStruct)
	sm.Child()
	sm.OnInt(5)
	assert.Equal(t, MatchState_Final, x._compartment_.State)
	assert.Equal(t, []string{"5"}, x.tape)

	sm = NewMatch()
	x = sm.(*matchStruct)
	sm.Child()
	sm.OnInt(3)
	assert.Equal(t, MatchState_ChildMatch, x._compartment_.State)
	assert.Equal(t, []string{"3", "?"}, x.tape)

	x.tape = x.tape[:0]
	sm.OnInt(42)
	assert.Equal(t, MatchState_ChildMatch, x._compartment_.State)
	assert.Equal(t, []string{"42 in child", "42"}, x.tape)

	x.tape = x.tape[:0]
	sm.OnInt(-200)
	assert.Equal(t, MatchState_ChildMatch, x._compartment_.State)
	assert.Equal(t, []string{"no match in child", "-200"}, x.tape)

	x.tape = x.tape[:0]
	sm.OnInt(100)
	assert.Equal(t, MatchState_ChildMatch, x._compartment_.State)
	assert.Equal(t, []string{"no match in child", "?"}, x.tape)
}

/* Test hierarchical string matching. */
func TestStringHierarchicalMatch(t *testing.T) {
	sm := NewMatch()
	x := sm.(*matchStruct)
	sm.Child()
	sm.Onstring("goodbye")
	assert.Equal(t, MatchState_Final, x._compartment_.State)
	assert.Empty(t, x.tape)

	sm = NewMatch()
	x = sm.(*matchStruct)
	sm.Child()
	sm.Onstring("hello")
	assert.Equal(t, MatchState_ChildMatch, x._compartment_.State)
	assert.Equal(t, []string{"hello in child", "hello"}, x.tape)

	x.tape = x.tape[:0]
	sm.Onstring("Testing 1, 2, 3...")
	assert.Equal(t, MatchState_ChildMatch, x._compartment_.State)
	assert.Equal(t, []string{"testing in child"}, x.tape)

	x.tape = x.tape[:0]
	sm.Onstring("$10!")
	assert.Equal(t, MatchState_ChildMatch, x._compartment_.State)
	assert.Equal(t, []string{"no match in child", "money"}, x.tape)

	x.tape = x.tape[:0]
	sm.Onstring("testing 1, 2, 3...")
	assert.Equal(t, MatchState_ChildMatch, x._compartment_.State)
	assert.Equal(t, []string{"no match in child", "?"}, x.tape)
}

File: framec_tests/java/Match/Match_test.java
package framec_tests.java.Match;

import java.util.*;
import static org.junit.Assert.assertEquals;
import org.junit.Test;

class MatchController extends Match {

    MatchController() {
      super();
    }

    protected void log_do(String msg) {
        this.tape.add(msg);
    }
}
// Test matching the empty string.
// TODO: Matching the empty string currently only works in multi-string
// patterns. The pattern `//`, which should match only the empty string,
// instead produces a parse error.
public class Match_test {
  @Test
  public void TestEmptyString(){
    MatchController sm = new MatchController();
    sm.Empty();
    sm.OnString("");
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("empty");
    assertEquals(sm.tape, expected);
    sm.tape = new ArrayList<>();
    sm.OnString("hi");
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("?");
    assertEquals(sm.tape, expected1);
  }
  
  @Test
  public void TestIntegerMatch(){
    MatchController sm = new MatchController();
    sm.Simple();
    sm.OnInt(0);
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("0");
    assertEquals(sm.tape, expected);
    sm.tape = new ArrayList<>();
    sm.OnInt(42);
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("42");
    assertEquals(sm.tape, expected1);
    sm.tape = new ArrayList<>();
    sm.OnInt(-200);
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add("-200");
    assertEquals(sm.tape, expected2);
  }

  /*
	Test simple string matching.

TODO: Testing revealed some limitations:
  - Frame does not support UTF-8 graphemes larger than 1 byte, so we're
    restricted to ASCII.
  - Frame does not have a way to match the '/' or '|' characters,
    which are part of the matching syntax.
*/

  @Test
  public void TestStringMatch(){
    MatchController sm = new MatchController();
    sm.Simple();
    sm.OnString("hello");
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("hello");
    assertEquals(sm.tape, expected);
    sm.tape = new ArrayList<>();
    sm.OnString("goodbye");
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("goodbye");
    assertEquals(sm.tape, expected1);
    sm.tape = new ArrayList<>();
    sm.OnString("Testing 1, 2, 3...");
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add("testing");
    assertEquals(sm.tape, expected2);
    ArrayList<String> expected3 = new ArrayList<String>();
    expected3.add("testing");
    assertEquals(sm.tape, expected3);
    sm.tape = new ArrayList<>();
    sm.OnString("$10!");
    ArrayList<String> expected4 = new ArrayList<String>();
    expected4.add("money");
    assertEquals(sm.tape, expected4);
    ArrayList<String> expected5 = new ArrayList<String>();
    expected5.add("money");
    assertEquals(sm.tape, expected5);
    sm.tape = new ArrayList<>();
    sm.OnString("missing");
    ArrayList<String> expected6 = new ArrayList<String>();
    expected6.add("?");
    assertEquals(sm.tape, expected6);
    ArrayList<String> expected7 = new ArrayList<String>();
    expected7.add("?");
    assertEquals(sm.tape, expected7);
    sm.tape = new ArrayList<>();
    sm.OnString("Testing");
    ArrayList<String> expected8 = new ArrayList<String>();
    expected8.add("?");
    assertEquals(sm.tape, expected8);
    ArrayList<String> expected9 = new ArrayList<String>();
    expected9.add("?");
    assertEquals(sm.tape, expected9);
    sm.tape = new ArrayList<>();
    sm.OnString("");
    ArrayList<String> expected10 = new ArrayList<String>();
    expected10.add("?");
    assertEquals(sm.tape, expected10);
    ArrayList<String> expected11 = new ArrayList<String>();
    expected11.add("?");
    assertEquals(sm.tape, expected11);

  }
  /* Test the multiple match syntax for integers. */

  @Test
  public void TestIntegerMultiMatch(){
    MatchController sm = new MatchController();
    sm.Multi();
    sm.OnInt(3);
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("3|-7");
    assertEquals(sm.tape, expected);
    sm.tape = new ArrayList<>();
    sm.OnInt(-7);
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("3|-7");
    assertEquals(sm.tape, expected1);
    sm.tape = new ArrayList<>();
    sm.OnInt(-4);
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add("-4|5|6");
    assertEquals(sm.tape, expected2);
    sm.tape = new ArrayList<>();
    sm.OnInt(5);
    ArrayList<String> expected3 = new ArrayList<String>();
    expected3.add("-4|5|6");
    assertEquals(sm.tape, expected3);
    sm.tape = new ArrayList<>();
    sm.OnInt(6);
    ArrayList<String> expected4 = new ArrayList<String>();
    expected4.add("-4|5|6");
    assertEquals(sm.tape, expected4);
    sm.tape = new ArrayList<>();
    sm.OnInt(4);
    ArrayList<String> expected5 = new ArrayList<String>();
    expected5.add("?");
    assertEquals(sm.tape, expected5);
    sm.tape = new ArrayList<>();
    sm.OnInt(0);
    ArrayList<String> expected6 = new ArrayList<String>();
    expected6.add("?");
    assertEquals(sm.tape, expected6);
    sm.tape = new ArrayList<>();
  }

  // Test the multiple match syntax for integers. Also tests matching
// whitespace-only strings.

  @Test
  public void TestStringMultiMatch(){
    MatchController sm = new MatchController();
    sm.Multi();
    sm.OnString("$10");
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("symbols");
    assertEquals(sm.tape, expected);
    sm.tape = new ArrayList<>();
    sm.OnString("12.5%");
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("symbols");
    assertEquals(sm.tape, expected1);
    sm.tape = new ArrayList<>();
    sm.OnString("@#*!");
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add("symbols");
    assertEquals(sm.tape, expected2);
    sm.tape = new ArrayList<>();
    sm.OnString(" ");
    ArrayList<String> expected3 = new ArrayList<String>();
    expected3.add("whitespace");
    assertEquals(sm.tape, expected3);
    sm.tape = new ArrayList<>();
    sm.OnString(" ");
    ArrayList<String> expected4 = new ArrayList<String>();
    expected4.add("whitespace");
    assertEquals(sm.tape, expected4);
    sm.tape = new ArrayList<>();
    sm.OnString("\t");
    ArrayList<String> expected5 = new ArrayList<String>();
    expected5.add("whitespace");
    assertEquals(sm.tape, expected5);
    sm.tape = new ArrayList<>();
    sm.OnString("\n");
    ArrayList<String> expected6 = new ArrayList<String>();
    expected6.add("whitespace");
    assertEquals(sm.tape, expected6);
    sm.tape = new ArrayList<>();
    sm.OnString("10");
    ArrayList<String> expected7 = new ArrayList<String>();
    expected7.add("?");
    assertEquals(sm.tape, expected7);
    sm.tape = new ArrayList<>();
    sm.OnString("#");
    ArrayList<String> expected8 = new ArrayList<String>();
    expected8.add("?");
    assertEquals(sm.tape, expected8);
    sm.tape = new ArrayList<>();
    sm.OnString("   ");
    ArrayList<String> expected9 = new ArrayList<String>();
    expected9.add("?");
    assertEquals(sm.tape, expected9);
    sm.tape = new ArrayList<>();
    sm.OnString("");
    ArrayList<String> expected10 = new ArrayList<String>();
    expected10.add("?");
    assertEquals(sm.tape, expected10);
    sm.tape = new ArrayList<>();
  }
  /* Test nested integer matching. */

  @Test
  public void TestIntegerNestedMatch(){
    MatchController sm = new MatchController();
    sm.Nested();
    sm.OnInt(1);
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("1-3");
    expected.add("1");
    assertEquals(sm.tape, expected);
    sm.tape = new ArrayList<>();
    sm.OnInt(2);
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("1-3");
    expected1.add("2");
    assertEquals(sm.tape, expected1);
    sm.tape = new ArrayList<>();
    sm.OnInt(3);
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add("1-3");
    expected2.add("3");
    assertEquals(sm.tape, expected2);
    sm.tape = new ArrayList<>();
    sm.OnInt(4);
    ArrayList<String> expected3 = new ArrayList<String>();
    expected3.add("4-5");
    expected3.add("4");
    assertEquals(sm.tape, expected3);
    sm.tape = new ArrayList<>();
    sm.OnInt(5);
    ArrayList<String> expected4 = new ArrayList<String>();
    expected4.add("4-5");
    expected4.add("5");
    assertEquals(sm.tape, expected4);
    sm.tape = new ArrayList<>();
    sm.OnInt(10);
    ArrayList<String> expected5 = new ArrayList<String>();
    expected5.add("too big");
    assertEquals(sm.tape, expected5);
    sm.tape = new ArrayList<>();
    sm.OnInt(0);
    ArrayList<String> expected6 = new ArrayList<String>();
    expected6.add("too small");
    assertEquals(sm.tape, expected6);
    sm.tape = new ArrayList<>();
  }

  /* Test nested string matching. */
  @Test
  public void TestStringNestedMatch(){
    MatchController sm = new MatchController();
    sm.Nested();
    sm.OnString("hello");
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("greeting");
    expected.add("English");
    assertEquals(sm.tape, expected);
    sm.tape = new ArrayList<>();
    sm.OnString("hola");
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("greeting");
    expected1.add("Spanish");
    assertEquals(sm.tape, expected1);
    sm.tape = new ArrayList<>();
    sm.OnString("bonjour");
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add("greeting");
    expected2.add("French");
    assertEquals(sm.tape, expected2);
    sm.tape = new ArrayList<>();
    sm.OnString("goodbye");
    ArrayList<String> expected3 = new ArrayList<String>();
    expected3.add("farewell");
    expected3.add("English");
    assertEquals(sm.tape, expected3);
    sm.tape = new ArrayList<>();
    sm.OnString("adios");
    ArrayList<String> expected4 = new ArrayList<String>();
    expected4.add("farewell");
    expected4.add("Spanish");
    assertEquals(sm.tape, expected4);
    sm.tape = new ArrayList<>();
    sm.OnString("au revoir");
    ArrayList<String> expected5 = new ArrayList<String>();
    expected5.add("farewell");
    expected5.add("French");
    assertEquals(sm.tape, expected5);
    sm.tape = new ArrayList<>();
    sm.OnString("hallo");
    ArrayList<String> expected6 = new ArrayList<String>();
    expected6.add("?");
    assertEquals(sm.tape, expected6);
    sm.tape = new ArrayList<>();
    sm.OnString("ciao");
    ArrayList<String> expected7 = new ArrayList<String>();
    expected7.add("?");
    assertEquals(sm.tape, expected7);
    sm.tape = new ArrayList<>();
  }
  /* Test hierarchical integer matching. */

  @Test
  public void TestIntegerHierarchicalMatch(){
    MatchController sm = new MatchController();
    sm.Child();
    sm.OnInt(0);
    assertEquals(sm.state_info(), "6");
    sm.tape = new ArrayList<>();

    sm = new MatchController();
    sm.Child();
    sm.OnInt(4);
    assertEquals(sm.state_info(), "5");
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("4");
    assertEquals(sm.tape, expected);

    sm.tape = new ArrayList<String>();
    sm.OnInt(5);
    assertEquals(sm.state_info(), "6");
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("5");
    assertEquals(sm.tape, expected1);
    
    sm = new MatchController();
    sm.tape = new ArrayList<String>();
    sm.Child();
    sm.OnInt(3);
    assertEquals(sm.state_info(), "5");
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add("3");
    expected2.add("?");
    assertEquals(sm.tape, expected2);

    sm.tape = new ArrayList<String>();
    sm.OnInt(42);
    assertEquals(sm.state_info(), "5");
    ArrayList<String> expected3 = new ArrayList<String>();
    expected3.add("42 in child");
    expected3.add("42");
    assertEquals(sm.tape, expected3);

    sm.tape = new ArrayList<String>();
    sm.OnInt(-200);
    assertEquals(sm.state_info(), "5");
    ArrayList<String> expected4 = new ArrayList<String>();
    expected4.add("no match in child");
    expected4.add("-200");
    assertEquals(sm.tape, expected4);

    sm.tape = new ArrayList<String>();
    sm.OnInt(100);
    assertEquals(sm.state_info(), "5");
    ArrayList<String> expected5 = new ArrayList<String>();
    expected5.add("no match in child");
    expected5.add("?");
    assertEquals(sm.tape, expected5);
  }

  /* Test hierarchical string matching. */
  @Test
  public void TestStringHierarchicalMatch(){
    MatchController sm = new MatchController();
    sm.Child();
    sm.OnString("goodbye");
    assertEquals(sm.state_info(), "6");
    sm.tape = new ArrayList<>();

    sm = new MatchController();
    sm.Child();
    sm.OnString("hello");
    assertEquals(sm.state_info(), "5");
    ArrayList<String> expected = new ArrayList<String>();
    expected.add("hello in child");
    expected.add("hello");
    assertEquals(sm.tape, expected);

    sm.tape = new ArrayList<>();
    sm.OnString("Testing 1, 2, 3...");
    assertEquals(sm.state_info(), "5");
    ArrayList<String> expected1 = new ArrayList<String>();
    expected1.add("testing in child");
    assertEquals(sm.tape, expected1);

    sm.tape = new ArrayList<>();
    sm.OnString("$10!");
    assertEquals(sm.state_info(), "5");
    ArrayList<String> expected2 = new ArrayList<String>();
    expected2.add("no match in child");
    expected2.add("money");
    assertEquals(sm.tape, expected2);

    sm.tape = new ArrayList<>();
    sm.OnString("testing 1, 2, 3...");
    assertEquals(sm.state_info(), "5");
    ArrayList<String> expected3 = new ArrayList<String>();
    expected3.add("no match in child");
    expected3.add("?");
    assertEquals(sm.tape, expected3);
  }
}

